src/librustc/middle/traits/mod.rs

```diff
@@ -95,15 +96,27 @@ pub struct ObligationCause<'tcx> {
     pub code: ObligationCauseCode<'tcx>
 }
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub enum ObligationCauseCode<'tcx> {
     /// Not well classified or should be obvious from span.
     MiscObligation,
 
+    /// Obligation that triggers warning until RFC 1214 is fully in place.
+    RFC1214(Rc<ObligationCauseCode<'tcx>>),
```

Comment:
So I was doing some experimentation and I remembered that I intentionally put the (new-style) wf checking at the very end of type-checking. Among other things, this means that for the most the part these warnings do not affect caches or anything else, since you only start seeing obligations with cause RFC1214 at the very end of the process. However, there are *some* cases where we enforce stricter rules during the main phase of type checking, so I will have to see if I can trigger an ICE or other undesirable behavior by focusing on those cases (so far I have failed).